File: python/LinkedList/MergeTwoSortedLists.py
"""
21. Merge Two Sorted Lists
Difficulty: Easy
Topics: LinkedList, Recursion
"""
import ListNode

def mergeTwoLists(List1: ListNode, list2: ListNode) -> ListNode:
    if not list1:
        return list2
    if not list2:
        return list1
    
    cur = dummy = ListNode()

    while list1 and list2:
        if list1.val <= list2.val:
            cur.next = list1
            list1 = list1.next
        else:
            cur.next =list2
            list2 = list2.next
        cur = cur.next
    
    # Process the rest of the remaining list (l1 or l2)
    if list1 or list2:
        cur.next = list1 if list1 else list2

    return dummy.next

File: java/LinkedList/MergeTwoSortedLists.java
package java.LinkedList;

public class MergeTwoSortedLists {
    /**
     * 21. Merge Two Sorted Lists
     * Difficulty: Easy
     * Topics: LinkedList, Recursion
     */
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        ListNode falseHead = new ListNode(-1);
        ListNode result = falseHead;
        
        while (list1 != null && list2 != null) {
            
            if(list1.val > list2.val) {
                result.next  = list2;
                list2 = list2.next;
            } else {
                result.next = list1;
                list1 = list1.next;
            }
            
            result = result.next;
            
        }
        
        if (list1 == null) {
            result.next = list2;
        } else {
            result.next = list1;
        }
        
        return falseHead.next;
        
    }
}
